Monero: Add CAIP-10 spec (#86)
ChainAgnostic/namespaces: monero/caip10.md

---
namespace-identifier: monero-caip10
title: Monero - Addresses
author: silverpill (@silverpill)
discussions-to: https://github.com/ChainAgnostic/namespaces/issues/41
status: Draft
type: Informational
created: 2023-09-01
requires: ["CAIP-10"]
---

# CAIP-10

*For context, see the [CAIP-10][] specification.*

> 

Reviewer: ```suggestion
CAIP-10 defines an pattern for identifying addresses in a cryptographic system as heirarchical URIs.
```

> CAIP-10 defines an pattern for identifying addresses in a cryptographic system as heirarchical URIs.

Reviewer: To be totally honest, CAIP-10 is final but I feel like the semantics of "account" aren't really explicit enough; if I can get some UTXO people to help me write it I might try writing an `Informational` CAIP to caveat and nuance this mapping a little, which is why I'm being so stubborn here about what does and doesn't work in the equivalence!

CAIP-10 doesn't really say if by "account" it means:
1. a "unit of accounting" (as in balance, whether public or requiring a key to see),
1. "distinct unit of account" (as in, a balance independent from any other),
1. "address" (string expresses a stable, addressable identity), AND
1. "actor" (as in payer/payee in a ledger-like transaction system).

In EVM and similar "account-based" networks, each wallet's address or CAIP-10 represents exactly one of each.  But in a system where there are not 1:1 relationships between the four, it's not self-evident which interpretation best serves interoperability across systems.

I have been going on the assumption that # 3 above is the most important, because the other three options make far less sense reading them against the [## Semantics section of btc/caip10.md](https://namespaces.chainagnostic.org/bip122/caip10#rationale), which says P2SH addresses are "close enough" to the EVM account model and implies by omission that earlier P2PKH addresses are not.  Since a given wallet/privatekey can generate multiple P2SH addresses and these payscripts can impose conditions, timelocks, etc, none of the other 3 seems valid.  To further undermine my confidence here, though, the example given by that profile is not, in fact, a P2SH address, but a P2PKH one (it starts with 1 instead of 3 in base58 form).  Later address types (taproot, segwit) aren't addressed at all!

I guess what I'm saying is that I'm over-editing your first line because I'm trying to make your summary of CAIP-10 more explicit than CAIP-10 is 😅

> CAIP-10 defines an pattern for identifying addresses in a cryptographic system as heirarchical URIs.

Author: I think of accounts as boxes that hold certain amounts of money. In Ethereum-like systems the balance is continuous: you can subtract or add any amount. In UTXO-based systems the box actually contains discrete "notes": you can take one note out and put a smaller one back, for example. It's called "address" but it nevertheless works like an account.

Some Bitcoin and Monero software introduce another level of abstraction ("wallets" or "accounts") in attempt to simplify address management, but in my opinion it's outside of CAIP-10 scope.

So in most (if not all) cases 1 and 3 will be the same thing (I don't quite understand the difference between 1 and 2). 4 is largely depends on how account is used. Sometimes it represents a unique identity. Sometimes several accounts are linked the one person.

>which says P2SH addresses are "close enough" to the EVM account model and implies by omission that earlier P2PKH addresses are not.

IIRC both P2SH and P2PKH can have balances and therefore can be viewed as CAIP-10 accounts. Same for segwit, taproot and others.

> CAIP-10 defines an pattern for identifying addresses in a cryptographic system as heirarchical URIs.

Reviewer: I think the trick here is that wallets aggregate accounts. for example, wallet A has accounts X, Y, and Z in them. each have 10btc in them. A can send someone 30btc taking notes out of all three-- that's what I meant by "independent" in #2. they can *operate* independently and each have a different balance, yet a single key/actor controls all three so that "independence" is only contingent on the information assymetry between A and its counterparties who don't realize those independent accounts are really one actor and can function as one account at will 😄

anyways thanks for thinking thru this with me, I think there's no change worth making here I just wanted to explore the mental model a bit!

## Rationale

CAIP-10 defines a pattern for identifying an account in a blockchain. 

> Monero chains use UTXO model and [Monero addresses][] are functionally similar to accounts in other namespaces.

Reviewer: ```suggestion
Monero chains use the UTXO accounting model, making [Monero addresses][] functionally similar to other UTXO-based namespaces, in that CAIP-10 identifiers represent **payee identifiers** (of which a given actor has many) rather than stable **actor identifiers**.
```

Author: >CAIP-10 identifiers represent **payee identifiers** (of which a given actor has many) rather than stable **actor identifiers**

I think this is not correct. Each Monero address represents a separate identity. It has its own balance, it can be re-used, etc.

Reviewer: Hmmm, you make a good point, they are independent and stable identities and in that sense a good analogue to an EVM-style account-address; what I'm confused on is the _independent balance_ part, which is where I thought the analogy broke down (to the glee of privacy maxis who ridicule EVM 😅 ).  

As for `payee`, I was kind of leaning on my admittedly superficial understanding of BTC terminology, where P2PK, P2PKH, and P2SH addresses are all "payees" (legitimate targets of a txn) even though they're not all actors or accounts. 

Can you say more about the independent balance? I was under the impression that I send 10 coins each to 10 different subaddresses of an empty wallet, the raw address of that account now has 100 coins it can send on to some third party?  I.e., I thought the "balances" of subaddresses were independent (but partial) views of a single balance (in the underlying ledger/accounting sense),  am I wrong?

Author: For practical purposes, subaddresses are distinct "boxes" with separate balances. See this RPC endpoint for example: https://www.getmonero.org/resources/developer-guides/wallet-rpc.html#get_balance

I don't know enough to say how it works under the hood. Generally, addresses in UTXO-based systems are abstractions, and on the base level there are discrete "notes" which you can spend only if you know the secret.

## Semantics

Monero has several types of addresses: [standard address](https://monerodocs.org/public-address/standard-address/), [subaddress](https://monerodocs.org/public-address/subaddress/) and [integrated address](https://monerodocs.org/public-address/integrated-address/). 
Any address contains network and address type identifiers, public spend key, public view key, a checksum, and may contain additional data depending on the address type. 

> [Base58-monero][base58-monero] encoding is used to encode the address into a string.

Reviewer: ```suggestion
Note that the [Base58-monero][base58-monero] encoding algorithm is used to encode the address from a byte array into a string; it uses the same alphabet as [base58btc][] but should not be decoded using [base58btc][] libraries due to the additional padding step to normalize string length.
```

> CAIP-10 "accounts" should not be confused with [Monero accounts][], which are used in Monero wallets for grouping related addresses and transactions.

Reviewer: Do you mean that in a Monero wallet, *counterparties'* various addresses are grouped into a logical unit representing the actor?  This would align with other UTXO systems and my mental model, but making it painfully explicit is usually my goal here :D

Also, does the term also organize transactions on, say, block explorers or other ledger views? 
```suggestion
CAIP-10 "accounts" should not be confused with "[Monero accounts][]", a term used in Monero wallets for grouping related addresses and transactions by actor.
```

Author: >Do you mean that in a Monero wallet, counterparties' various addresses are grouped into a logical unit representing the actor?

No, there are no on-chain links between addresses. Accounts only exist in the wallet, mostly for convenience. There was a similar entity in Bitcoin software, although they removed it later (see this RPC method: https://bitcoincore.org/en/doc/0.16.0/rpc/wallet/getaccount/).

Overall, Monero addresses are very similar to Bitcoin and Ethereum addresses.

>Also, does the term also organize transactions on, say, block explorers or other ledger views?

No.

Author: >a term used in Monero wallets for grouping related addresses and transactions by actor.

I think accounts are usually used to categorize one's own finances.

Reviewer: Right, so it's a UX construct, not any kind of onchain data structure.  That makes sense-- definitely worth warning people new to Monero that they will see the term "account" used in this way and it will NOT be a 1:1 map to CAIP-10s!
